lib/elixir/lib/calendar/naive_datetime.ex

```diff
@@ -571,6 +571,73 @@ defmodule NaiveDateTime do
     units1 - units2
   end
 
+  @doc """
+  Shifts a naive datetime by given duration according to its calendar.
+
+  Check `Calendar.ISO.shift_naive_datetime/8` for more information.
+
+  ## Examples
+
+      iex> NaiveDateTime.shift(~N[2016-01-31 00:00:00], month: 1)
+      {:ok, ~N[2016-02-29 00:00:00]}
+      iex> NaiveDateTime.shift(~N[2016-01-31 00:00:00], year: 4, day: 1)
+      {:ok, ~N[2020-02-01 00:00:00]}
+      iex> NaiveDateTime.shift(~N[2016-01-31 00:00:00], second: 45)
+      {:ok, ~N[2016-01-31 00:00:45]}
+      iex> NaiveDateTime.shift(~N[2016-01-31 00:00:00], microsecond: 100)
+      {:ok, ~N[2016-01-31 00:00:00.000100]}
+
+  """
+  @spec shift(
+          Calendar.naive_datetime(),
+          Duration.t() | [Duration.unit()]
+        ) :: {:ok, t} | {:error, :invalid_duration}
+  def shift(%NaiveDateTime{calendar: calendar} = naive_datetime, %Duration{} = duration) do
+    %{
+      year: year,
+      month: month,
+      day: day,
+      hour: hour,
+      minute: minute,
+      second: second,
+      microsecond: microsecond
+    } = naive_datetime
+
+    {year, month, day, hour, minute, second, microsecond} =
+      calendar.shift_naive_datetime(
+        year,
+        month,
+        day,
+        hour,
+        minute,
+        second,
+        microsecond,
+        duration
+      )
+
+    {:ok,
+     %NaiveDateTime{
+       calendar: calendar,
+       year: year,
+       month: month,
+       day: day,
+       hour: hour,
+       minute: minute,
+       second: second,
+       microsecond: microsecond
+     }}
+  end
+
+  def shift(%NaiveDateTime{} = naive_datetime, duration_units) do
+    case Duration.new(duration_units) do
+      {:ok, duration} ->
+        shift(naive_datetime, duration)
+
+      {:error, :invalid_duration} ->
+        {:error, :invalid_duration}
```

Comment:
No error here either. We can just let Duration.new raise.

Reply:
👍 i'll update all the calendar types to use `Duration.new!/1`

Reply:
Sorry, to be clear: there is no `Duration.new!`. Only `Duration.new` and it raises on invalid options.